### PR TITLE
add sale channel listing

### PR DESCRIPTION
**Pull request by d-wysocki**
I want to merge this change because...I want do add SaleChannelListing 

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/discount/models.py [hunk 1]
```diff
@@ -225,24 +221,50 @@ class Meta:
...
             return partial(fixed_discount, discount=discount_amount)
         if self.type == DiscountValueType.PERCENTAGE:
-            return partial(percentage_discount, percentage=self.value)
+            return partial(
+                percentage_discount, percentage=sale_channel_listing.discount_value,
+            )
         raise NotImplementedError("Unknown discount type")
 
 
+class SaleChannelListing(models.Model):
```
Comment (fowczarek) [inline saleor/discount/models.py:245]: We should add `unique_together`

File: saleor/discount/utils.py [hunk 2]
```diff
@@ -49,7 +49,7 @@ def remove_voucher_usage_by_customer(voucher: "Voucher", customer_email: str) ->
 
 
 def get_product_discount_on_sale(
-    product: "Product", product_collections: Set[int], discount: DiscountInfo
+    product: "Product", product_collections: Set[int], discount: DiscountInfo, channel
```
Comment (fowczarek) [inline saleor/discount/utils.py:52]: We should add typing.

File: saleor/discount/utils.py [hunk 3]
```diff
@@ -58,21 +58,25 @@ def get_product_discount_on_sale(
...
     raise NotApplicable("Discount not applicable for this product")
 
 
 def get_product_discounts(
     *,
     product: "Product",
     collections: Iterable["Collection"],
-    discounts: Iterable[DiscountInfo]
+    discounts: Iterable[DiscountInfo],
+    channel
```
Comment (fowczarek) [inline saleor/discount/utils.py:71]: We should add typing.

File: saleor/discount/utils.py [hunk 4]
```diff
@@ -82,13 +86,17 @@ def calculate_discounted_price(
     product: "Product",
     price: Money,
     collections: Iterable["Collection"],
-    discounts: Optional[Iterable[DiscountInfo]]
+    discounts: Optional[Iterable[DiscountInfo]],
+    channel
```
Comment (fowczarek) [inline saleor/discount/utils.py:90]: We should add typing.

File: saleor/graphql/discount/mutations.py [hunk 5]
```diff
@@ -287,16 +299,25 @@ class SaleInput(graphene.InputObjectType):
...
     def success_response(cls, instance):
         # Update the "discounted_prices" of the associated, discounted
         # products (including collections and categories).
         update_products_discounted_prices_of_discount_task.delay(instance.pk)
         return super().success_response(instance)
 
+    @classmethod
+    def perform_mutation(cls, _root, info, **data):
+        response = super().perform_mutation(_root, info, **data)
+        instance = ChannelContext(node=response.sale, channel_slug=None)
```
Comment (fowczarek) [inline saleor/graphql/discount/mutations.py:316]: Maby better to put this line to `success_response`?

File: saleor/graphql/discount/mutations.py [hunk 6]
```diff
@@ -382,4 +403,83 @@ def perform_mutation(cls, _root, info, **data):
             info, data.get("id"), only_type=Sale, field="sale_id"
         )
         cls.remove_catalogues_from_node(sale, data.get("input"))
-        return SaleRemoveCatalogues(sale=sale)
+        return SaleRemoveCatalogues(sale=ChannelContext(node=sale, channel_slug=None))
+
+
+class SaleChannelListingAddInput(graphene.InputObjectType):
+    channel_id = graphene.ID(required=True, description="ID of a channel.")
+    discount_value = PositiveDecimal(description="The value of the discount.")
```
Comment (fowczarek) [inline saleor/graphql/discount/mutations.py:411]: I think this vale should be required.
Whats means if the user wants to create/update `SaleChannelListing` in the channel without value?

File: saleor/graphql/discount/mutations.py [hunk 7]
```diff
@@ -382,4 +403,83 @@ def perform_mutation(cls, _root, info, **data):
...
+        error_type_class = DiscountError
+        error_type_field = "discount_errors"
+
+    @classmethod
+    def add_channels(cls, sale: "SaleModel", add_channels: List[Dict]):
+        for add_channel in add_channels:
+            defaults = {}
+            discount_value = add_channel.get("discount_value")
+            if discount_value:
+                defaults["discount_value"] = discount_value
```
Comment (fowczarek) [inline saleor/graphql/discount/mutations.py:449]: Lack of `validate_price_precision` and test for it. 
This validation is always required for any prices form inputs.

File: saleor/graphql/discount/mutations.py [hunk 8]
```diff
@@ -382,4 +403,83 @@ def perform_mutation(cls, _root, info, **data):
...
+    def perform_mutation(cls, _root, info, id, input):
+        sale = cls.get_node_or_error(info, id, only_type=Sale, field="id")
+        errors = defaultdict(list)
+
+        clean_channels = cls.clean_channels(
+            info, input, errors, DiscountErrorCode.DUPLICATED_INPUT_ITEM.value
+        )
+
+        if errors:
+
```
Comment (fowczarek) [inline saleor/graphql/discount/mutations.py:479]: Maby?
```suggestion
```

File: saleor/graphql/discount/schema.py [hunk 9]
```diff
@@ -95,3 +112,4 @@ class DiscountMutations(graphene.ObjectType):
     voucher_catalogues_add = VoucherAddCatalogues.Field()
     voucher_catalogues_remove = VoucherRemoveCatalogues.Field()
     voucher_translate = VoucherTranslate.Field()
+    sale_channel_listing_update = SaleChannelListingUpdate.Field()
```
Comment (fowczarek) [inline saleor/graphql/discount/schema.py:115]: Visual issue :police_officer:

File: saleor/discount/models.py [hunk 10]
```diff
@@ -225,24 +221,50 @@ class Meta:
...
-            self.get_type_display(),
-        )
+        return "Sale(name=%r, type=%s)" % (str(self.name), self.get_type_display(),)
 
     def __str__(self):
         return self.name
 
-    def get_discount(self):
+    def get_discount(self, channel: Channel):
+        sale_channel_listing = self.channel_listing.filter(channel=channel).first()
```
Comment (maarcingebala) [inline saleor/discount/models.py:230]: Making database queries in model functions is a bad idea as it often leads to unwanted duplicated DB queries. We should either refactor this function to move it outside of the model or pass sale_channel_listing as an argument.

File: saleor/graphql/discount/mutations.py (hunk 8 above)
Comment (maarcingebala) [reply saleor/graphql/discount/mutations.py:479]: Agree, in short and simple functions too many additional empty lines doesn't help much with readability.

File: saleor/graphql/discount/types.py [hunk 11]
```diff
@@ -21,28 +36,58 @@ class Sale(CountableDjangoObjectType):
...
     @staticmethod
-    def resolve_products(root: models.Sale, info, **_kwargs):
-        return root.products.visible_to_user(info.context.user)
+    def resolve_products(root: ChannelContext[models.Sale], info, **_kwargs):
+        return root.node.products.visible_to_user(info.context.user)
+
+    @staticmethod
+    def resolve_discount_value(root: ChannelContext[models.Sale], *_args, **_kwargs):
+        channel_listing = root.node.channel_listing.filter(
+            channel__slug=str(root.channel_slug)
```
Comment (maarcingebala) [inline saleor/graphql/discount/types.py:81]: I think `str()` is not needed here?

File: saleor/graphql/discount/types.py [hunk 12]
```diff
@@ -21,28 +36,58 @@ class Sale(CountableDjangoObjectType):
...
+        ).first()
+        return channel_listing.discount_value if channel_listing else None
+
+    @staticmethod
+    def resolve_currency(root: ChannelContext[models.Sale], *_args, **_kwargs):
+        channel_listing = root.node.channel_listing.filter(
+            channel__slug=str(root.channel_slug)
```
Comment (maarcingebala) [inline saleor/graphql/discount/types.py:88]: And here.

File: saleor/graphql/discount/tests/test_bulk_delete.py [hunk 13]
```diff
@@ -1,16 +1,23 @@
...
-def sale_list():
-    sale_1 = Sale.objects.create(name="Sale 1", value=5)
-    sale_2 = Sale.objects.create(name="Sale 2", value=5)
-    sale_3 = Sale.objects.create(name="Sale 3", value=5)
-    return sale_1, sale_2, sale_3
+def sale_list(channel_USD):
+    sale_1 = Sale.objects.create(name="Sale 1")
+    sale_2 = Sale.objects.create(name="Sale 2")
+    sale_3 = Sale.objects.create(name="Sale 3")
+    sales = [sale_1, sale_2, sale_3]
```
Comment (fowczarek) [inline saleor/graphql/discount/tests/test_bulk_delete.py:13]:
```suggestion
    sales = Sale.objects.bulk_create([Sale(name="Sale 1"), Sale(name="Sale 2"), Sale(name="Sale 3")])
```

File: saleor/graphql/discount/types.py (hunk 11 above)
Comment (fowczarek) [reply saleor/graphql/discount/types.py:81]: It's needed because chanel_slug  can be `SimpleLasyObject` if someone doesn't pass chanel_slug in query and ask for `discount_value`.
```
sales(first:10){
    name,
    discountValue
}
```

File: saleor/graphql/discount/types.py (hunk 12 above)
Comment (fowczarek) [reply saleor/graphql/discount/types.py:88]: Same as above.

File: saleor/graphql/discount/tests/test_sale_channel_listing.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
+from unittest.mock import patch
+
+import graphene
+
+from ....discount.models import SaleChannelListing
+from ...tests.utils import assert_negative_positive_decimal_value, get_graphql_content
+
+SALE_CHANNEL_LISTING_UPDATE_MUTATION = """
```
Comment (fowczarek) [inline saleor/graphql/discount/tests/test_sale_channel_listing.py:8]: Lack of test for :
- duplicated IDs in add 
- duplicated IDs in remove 
- Same id in add and remove
- invalid decimal precision

Comment (fowczarek) [reply saleor/graphql/discount/tests/test_sale_channel_listing.py:8]: In these tests, we should add custom param to errors to identify witch input course error.

File: saleor/graphql/discount/types.py [hunk 15]
```diff
@@ -1,17 +1,32 @@
...
-class Sale(CountableDjangoObjectType):
+class SaleChannelListing(CountableDjangoObjectType):
+    class Meta:
+        description = "Represents sale channel listing."
+        model = models.SaleChannelListing
+        interfaces = [relay.Node]
+        only_fields = [
+            "id",
+            "channel",
+            "discount_value",
```
Comment (fowczarek) [inline saleor/graphql/discount/types.py:25]: Lack of `currency`

File: saleor/graphql/discount/types.py [hunk 16]
```diff
@@ -21,28 +36,58 @@ class Sale(CountableDjangoObjectType):
...
     @staticmethod
-    def resolve_categories(root: models.Sale, *_args, **_kwargs):
-        return root.categories.all()
+    def resolve_categories(root: ChannelContext[models.Sale], *_args, **_kwargs):
+        return root.node.categories.all()
+
+    @staticmethod
+    @permission_required(DiscountPermissions.MANAGE_DISCOUNTS)
+    def resolve_channel_listing(root: ChannelContext[models.Sale], *_args, **_kwargs):
+        return root.node.channel_listing.all()
```
Comment (fowczarek) [inline saleor/graphql/discount/types.py:68]:
```suggestion
    def resolve_channel_listing(root: ChannelContext[models.Sale], *_args, **_kwargs):
        # TODO: Add dataloader.
        return root.node.channel_listing.all()
```

File: saleor/graphql/discount/types.py [hunk 17]
```diff
@@ -21,28 +36,58 @@ class Sale(CountableDjangoObjectType):
...
 
     @staticmethod
     def resolve_collections(root: models.Sale, info, **_kwargs):
```
Comment (fowczarek) [inline saleor/graphql/discount/types.py:71]: ChannelContext? 
If you don't find it we should add a test for query  `collections`  from Sale.

File: saleor/graphql/product/tests/benchmark/conftest.py [hunk 18]
```diff
@@ -1,16 +1,20 @@
...
-    return list(
-        Sale.objects.bulk_create(
-            [Sale(name="Sale1", value=15), Sale(name="Sale2", value=5)]
-        )
+def sales_list(channel_USD):
+    sales = Sale.objects.bulk_create([Sale(name="Sale1"), Sale(name="Sale2")])
+    values = [15, 5]
+    SaleChannelListing.objects.bulk_create(
+        [
+            SaleChannelListing(sale=sale, channel=channel_USD, discount_value=values[i])
```
Comment (fowczarek) [inline saleor/graphql/product/tests/benchmark/conftest.py:13]: Lack of `currency`

File: saleor/tests/fixtures.py [hunk 19]
```diff
@@ -1892,8 +1893,9 @@ def dummy_payment_data(payment_dummy):
 
 
 @pytest.fixture
-def sale(product, category, collection):
-    sale = Sale.objects.create(name="Sale", value=5)
+def sale(product, category, collection, channel_USD):
+    sale = Sale.objects.create(name="Sale")
+    SaleChannelListing.objects.create(sale=sale, channel=channel_USD, discount_value=5)
```
Comment (fowczarek) [inline saleor/tests/fixtures.py:1898]: Lack of `currency`

**Review by fowczarek — COMMENTED** [review]
I don't see test for

File: saleor/discount/utils.py [hunk 20]
```diff
@@ -13,7 +13,7 @@
 
 if TYPE_CHECKING:
     # flake8: noqa
-    from .models import Voucher
+    from .models import Voucher, Channel
```
Comment (fowczarek) [inline saleor/discount/utils.py:16]: Channels are not in `saleor.discount.models`.

File: saleor/discount/utils.py [hunk 21]
```diff
@@ -58,21 +61,28 @@ def get_product_discount_on_sale(
         or product_collections.intersection(discount.collection_ids)
     )
     if is_product_on_sale:
-        return discount.sale.get_discount()
+        sale_channel_listing = discount.sale.channel_listing.filter(  # type: ignore
+            channel=channel
+        ).first()
```
Comment (fowczarek) [inline saleor/discount/utils.py:66]: Maby we we should add `sale_channel_listing` to `DiscountInfo`?
In the current implementation, the query will be executed for each product separately.

File: saleor/graphql/discount/mutations.py [hunk 22]
```diff
@@ -366,7 +380,7 @@ def perform_mutation(cls, _root, info, **data):
             info, data.get("id"), only_type=Sale, field="sale_id"
         )
         cls.add_catalogues_to_node(sale, data.get("input"))
-        return SaleAddCatalogues(sale=sale)
+        return SaleAddCatalogues(sale=ChannelContext(node=sale, channel_slug=None))
```
Comment (fowczarek) [inline saleor/graphql/discount/mutations.py:383]: Can we restore the old return? 
We adding ChannelContext in `succesResponse` in `SaleUpdateDiscountedPriceMixin`

File: saleor/graphql/discount/mutations.py [hunk 23]
```diff
@@ -382,4 +396,117 @@ def perform_mutation(cls, _root, info, **data):
...
+            currency_code = channel.currency_code
+            if discount_value:
+                defaults["discount_value"] = discount_value
+            SaleChannelListing.objects.update_or_create(
+                sale=sale, channel=channel, currency=currency_code, defaults=defaults,
+            )
+
+    @classmethod
+    def clean_prices(cls, cleaned_channels, sale_type, errors, error_code):
+        if sale_type != DiscountValueType.FIXED:
```
Comment (fowczarek) [inline saleor/graphql/discount/mutations.py:453]: Can we add a test for `Percentage` discount we currency with less than 3 decimal precision and value for example 1.005?

File: saleor/graphql/discount/mutations.py [hunk 24]
```diff
@@ -382,4 +396,117 @@ def perform_mutation(cls, _root, info, **data):
...
+            ]
+            errors["input"].append(
+                ValidationError(
+                    error_message,
+                    code=error_code,
+                    params={"channels": list(invalid_channels_ids)},
+                )
+            )
+        else:
+            return cleaned_channels
```
Comment (fowczarek) [inline saleor/graphql/discount/mutations.py:482]:
```suggestion
        return cleaned_channels
```

File: saleor/graphql/discount/tests/test_pagination.py [hunk 25]
```diff
@@ -5,47 +5,49 @@
...
 from ....discount import DiscountValueType, VoucherType
-from ....discount.models import Sale, Voucher
+from ....discount.models import Sale, SaleChannelListing, Voucher
 from ...tests.utils import get_graphql_content
 
 
 @pytest.fixture
 @freeze_time("2020-03-18 12:00:00")
-def sales_for_pagination(db):
+def sales_for_pagination(db, channel_USD):
```
Comment (fowczarek) [inline saleor/graphql/discount/tests/test_pagination.py:14]: If we add fixture witch use db we can delete db form params. 
```suggestion
def sales_for_pagination(channel_USD):
```

File: saleor/graphql/discount/mutations.py (hunk 22 above)
Comment (d-wysocki) [reply saleor/graphql/discount/mutations.py:383]: `SaleAddCatalogues ` and `SaleRemoveCatalogues` itherit from `BaseDiscountCatalogueMutation(BaseMutation)`, we can use `success_response ` only with `ModelMutation`.

`SaleUpdateDiscountedPriceMixin ` is related with other mutations:

- SaleCreate

- SaleUpdate

- SaleDelete

File: saleor/graphql/discount/mutations.py (hunk 23 above)
Comment (d-wysocki) [reply saleor/graphql/discount/mutations.py:453]: after a conversation with @fowczarek, we will skip this test for now.

**Review by fowczarek — APPROVED** [review]

File: saleor/settings.py [hunk 26]
```diff
@@ -506,6 +506,7 @@ def get_host():
         "saleor.graphql.middleware.OpentracingGrapheneMiddleware",
         "saleor.graphql.middleware.JWTMiddleware",
         "saleor.graphql.middleware.app_middleware",
+        "graphene_django.debug.DjangoDebugMiddleware",
```
Comment (maarcingebala) [inline saleor/settings.py:509]: Is this related to this PR?

**Review by maarcingebala — APPROVED** [review]

File: saleor/settings.py (hunk 26 above)
Comment (fowczarek) [reply saleor/settings.py]: We should remove it. It's a legacy of optimization.

Comment (d-wysocki) [reply saleor/settings.py]: I have deleted it. Good catch, thanks :)
